clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java

```diff
@@ -63,69 +64,74 @@ public void configure(Map<String, ?> configs) {}
     @Override
     public void init(List<KafkaMetric> metrics) {
         synchronized (LOCK) {
-            for (KafkaMetric metric : metrics)
-                addAttribute(metric);
-            for (KafkaMbean mbean : mbeans.values())
+            for (KafkaMetric metric : metrics) {
+                final KafkaMbean mbean = addAttribute(metric);
+                if (!this.mbeans.containsKey(mbean.metricName)) {
+                    this.mbeans.put(mbean.metricName, mbean);
+                }
+            }
+            for (KafkaMbean mbean : this.mbeans.values()) {
                 reregister(mbean);
+            }
         }
     }
 
     @Override
     public void metricChange(KafkaMetric metric) {
         synchronized (LOCK) {
-            KafkaMbean mbean = addAttribute(metric);
+            final KafkaMbean mbean = addAttribute(metric);
+            if (!this.mbeans.containsKey(mbean.metricName)) {
+                this.mbeans.put(mbean.metricName, mbean);
+            }
             reregister(mbean);
         }
     }
 
     @Override
     public void metricRemoval(KafkaMetric metric) {
         synchronized (LOCK) {
-            KafkaMbean mbean = removeAttribute(metric);
-            if (mbean != null) {
-                if (mbean.metrics.isEmpty())
-                    unregister(mbean);
-                else
-                    reregister(mbean);
+            final KafkaMbean mbean = removeAttribute(metric);
+            if (mbean.metrics.isEmpty()) {
+                unregister(mbean);
+                this.mbeans.remove(mbean.metricName);
+            } else {
+                reregister(mbean);
             }
         }
     }
 
     private KafkaMbean removeAttribute(KafkaMetric metric) {
-        MetricName metricName = metric.metricName();
-        String mBeanName = getMBeanName(metricName);
-        KafkaMbean mbean = this.mbeans.get(mBeanName);
-        if (mbean != null)
+        final MetricName metricName = metric.metricName();
+        final KafkaMbean mbean = getMBean(metricName);
```

Comment:
I know... that's a questionable improvement, I agree. The idea was to use a common accessor in all the places. If you think that overhead is too much, please inform me - I'll revert the change.